File: Frontend/trackr-administrador/src/app/shared/http/examen/nivel-examen.service.ts
import { HttpClient } from '@angular/common/http';
import { Injectable } from '@angular/core';
import { NivelExamen } from '@models/examen/nivel-examen';
import { Observable } from 'rxjs';

@Injectable()
export class NivelExamenService {
  private dataUrl = 'nivelExamen/';

  constructor(public http: HttpClient) {}

  public consultarTodosParaSelector(): Observable<NivelExamen[]> {
    return this.http.get<NivelExamen[]>(this.dataUrl + 'consultarTodosParaSelector');
  }

  public consultar(idNivelExamen: number): Observable<NivelExamen> {
    return this.http.get<NivelExamen>(this.dataUrl + `consultar/${idNivelExamen}`);
  }

  public consultarGeneral(): Observable<NivelExamen[]> {
    return this.http.get<NivelExamen[]>(this.dataUrl + 'consultarGeneral');
  }

  public agregar(nivelExamen: NivelExamen): Observable<number> {
    return this.http.post<number>(this.dataUrl + 'agregar', nivelExamen);
  }

  public editar(nivelExamen: NivelExamen): Observable<void> {
    return this.http.put<void>(this.dataUrl + 'editar', nivelExamen);
  }

  public eliminar(idNivelExamen: number): Observable<void> {
    return this.http.delete<void>(this.dataUrl + `eliminar/${idNivelExamen}`);
  }
}

File: Backend/TrackrAPI/TrackrAPI/Services/Examen/NivelExamenService.cs
using TrackrAPI.Dtos.Examen;
using TrackrAPI.Models;
using TrackrAPI.Repositorys.Examen;

namespace TrackrAPI.Services.Examen;

public class NivelExamenService
{
    private readonly INivelExamenRepository _nivelExamenRepository;
    private readonly NivelExamenValidatorService _nivelExamenValidatorService;

    public NivelExamenService(
        INivelExamenRepository nivelExamenRepository,
        NivelExamenValidatorService nivelExamenValidatorService)
    {
        _nivelExamenRepository = nivelExamenRepository;
        _nivelExamenValidatorService = nivelExamenValidatorService;
    }

    public NivelExamenDto? Consultar(int idNivelExamen)
    {
        return _nivelExamenRepository.Consultar(idNivelExamen);
    }

    public IEnumerable<NivelExamenGridDto> ConsultarGeneral()
    {
        return _nivelExamenRepository.ConsultarGeneral();
    }

    public IEnumerable<NivelExamenGridDto> ConsultarTodosParaSelector()
    {
        return _nivelExamenRepository.ConsultarTodosParaSelector();
    }

    public int Agregar(NivelExamen nivelExamen)
    {
        _nivelExamenValidatorService.ValidarAgregar(nivelExamen);
        _nivelExamenValidatorService.ValidarDuplicado(nivelExamen);
        _nivelExamenRepository.Agregar(nivelExamen);
        return nivelExamen.IdNivelExamen;
    }

    public void Editar(NivelExamen nivelExamen)
    {
        _nivelExamenValidatorService.ValidarEditar(nivelExamen);
        _nivelExamenValidatorService.ValidarDuplicado(nivelExamen);
        _nivelExamenRepository.Editar(nivelExamen);
    }

    public void Eliminar(int idNivelExamen)
    {
        NivelExamenDto? nivelExamen = _nivelExamenRepository.Consultar(idNivelExamen);
        _nivelExamenValidatorService.ValidarEliminar(idNivelExamen);

        if (nivelExamen != null)
        {
            NivelExamen nivelExamenMod = new()
            {
                IdNivelExamen = nivelExamen.IdNivelExamen,
                Clave = nivelExamen.Clave,
                Descripcion = nivelExamen.Descripcion,
                FechaAlta = nivelExamen.FechaAlta,
                Estatus = false
            };

            _nivelExamenRepository.Editar(nivelExamenMod);
        }
    }
}
